### DISCUSSION TITLE
How to serve Apis Routes in the context of test execution with Testing Library?

### DISCUSSION THREAD

**Discussion by xereda** [vercel/next.js#46703] [Help]
### Summary

I use Next API Routes as a BFF server for the front-end application. I'm currently writing unit testing for the APIs, but I'd like to write more effective integration tests.

For that, I'm making a POC, where I intend to make mocks only for the original APIs (backend APIs) and no longer serve mocks of the Next.js Api Routes.

This would cause the processes contained in the Next APIs to be indirectly tested with the Testing Library and component tests. For example, if an API Route, as BFF responsibility, transforms a number into a string in a monetary format, the validation of this transformation would be directly contemplated by an integration test and the Testing Library.

However, I am having a problem. How to serve the Routes APIs together with the Testing Library execution?

Thanks in advance for your help!

**Comment by icyJoseph** [comment]
I guess you could... use MSW? https://mswjs.io/

Whenever your client code makes a call, if you setup an MSW listener, it'll catch the request, and then you can feed that into your API route. It is a bit hacky, but it would work, you just need to couple the MSW req/res objects with what Next.js expects. Not so trivial, but doable.

You could also just return the mocked data through MSW, but that's no what you want. Or maybe this one? https://www.npmjs.com/package/next-test-api-route-handler

Sorry for throwing a bunch of libraries at you, but maintaining these in a codebase is not very practical.

**Comment by xereda** [comment]
Hi @icyJoseph Thank you for the feedback! I'm using MSW. It's incredible tool! But, I was trying not to use it for the next apis route. However, adopting this practice, I believe that I'll problems in the future.

I will continue to write unit tests for my APIs (/api/*) and it is working great.
